src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,209 @@
 package se.liu.ida.hefquin.engine.datastructures.impl;
 
-import java.util.Iterator;
+import java.util.*;
 
 import org.apache.jena.graph.Node;
 import org.apache.jena.sparql.core.Var;
 
+import org.apache.jena.sparql.engine.binding.Binding;
 import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
 
 public class SolutionMappingsHashTable extends SolutionMappingsIndexBase
 {
+	// Having List<Node> as key type for the hash table is probably
+	// not the best choice in terms of efficiency. However, it will have
+	// to do for the moment.
+	// TODO: can this be made more efficient?
+	protected final Map<List<Node>, List<SolutionMapping>> map = new HashMap<>();
+	protected final Collection<Var> joinVariables;
+
+	public SolutionMappingsHashTable( final Set<Var> joinVariables ){
+		this.joinVariables = joinVariables;
+	}
+	public SolutionMappingsHashTable( final Var ... vars ) {
+		this.joinVariables = Arrays.asList(vars);
+	}
+
 	@Override
 	public int size() {
-		// TODO Auto-generated method stub
-		return 0;
+		int size = 0;
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while( li.hasNext() ){
+			size = size + li.next().size();
+		}
+		return size;
 	}
 
 	@Override
 	public boolean isEmpty() {
-		// TODO Auto-generated method stub
-		return false;
+		return ! ( !map.isEmpty() && size()>0 );
 	}
 
 	@Override
 	public boolean contains( final Object o ) {
-		// TODO Auto-generated method stub
-		return false;
+		if( !(o instanceof SolutionMapping) ){
+			throw new IllegalArgumentException();
+		}
+
+		boolean contain = false;
+		if( !map.isEmpty() ){
+			final Iterator<List<SolutionMapping>> li = map.values().iterator();
+			while( li.hasNext() ){
+				if( li.next().contains((SolutionMapping) o) ){
+					contain = true;
+				}
+			}
+		}
+		return contain;
```

Comment:
This can be simplified in several ways.

First, there is no need for the `contain` variable. Instead, simply return `true` inside the loop as soon as the solution mapping is found in a list. No need to continue looking into the rest of the lists!

Second, there is no need for the `if` statement. If `map` is empty, then `map.values()` will be empty and, thus, the loop will not even be entered.

Third, instead of using the iterator, you can use a `for` loop.

Fourth, casting `o` to a `SolutionMapping` should be done before the loop. Otherwise, it is done over and over again.

Hence, overall, you can replace this whole part as follows.
```suggestion
		final SolutionMapping sm = (SolutionMapping) o;
		for ( final List<SolutionMapping> li : map.values() ) {
			if ( l.contains(sm) ) {
				return true;
			}
		}
		return false;
```